supertokens_python/recipe/multifactorauth/utils.py

```diff
@@ -0,0 +1,221 @@
+                        tenant_id, FactorIds.THIRDPARTY, user_context
+                    )
+                    if valid_res.status == "TENANT_NOT_FOUND_ERROR":
+                        raise SuperTokensSessionError(
+                            SuperTokensSessionError.UNAUTHORISED, "Tenant not found"
+                        )
+                    elif valid_res.status == "OK":
+                        computed_first_factor_id_for_session = FactorIds.THIRDPARTY
+                        break
+                else:
+                    factors_to_check = []
+                    if login_method.email is not None:
+                        factors_to_check.extend(
+                            [FactorIds.LINK_EMAIL, FactorIds.OTP_EMAIL]
+                        )
+                    if login_method.phone_number is not None:
+                        factors_to_check.extend(
+                            [FactorIds.LINK_PHONE, FactorIds.OTP_PHONE]
+                        )
+
+                    for factor_id in factors_to_check:
+                        valid_res = await is_valid_first_factor(
+                            tenant_id, factor_id, user_context
+                        )
+                        if valid_res.status == "TENANT_NOT_FOUND_ERROR":
+                            raise SuperTokensSessionError(
+                                SuperTokensSessionError.UNAUTHORISED, "Tenant not found"
+                            )
+                        elif valid_res.status == "OK":
+                            computed_first_factor_id_for_session = factor_id
+                            break
+
+                    if computed_first_factor_id_for_session is not None:
+                        break
+
+        if computed_first_factor_id_for_session is None:
+            raise SuperTokensSessionError(
+                SuperTokensSessionError.UNAUTHORISED, "Incorrect login method used"
+            )
+
+        updated_claim_val = True
+        mfa_claim_value = {
+            "c": {computed_first_factor_id_for_session: first_factor_time},
+            "v": True,  # updated later in this function
+        }
+
+    completed_factors = mfa_claim_value["c"]
+
+    user_prom: Optional[asyncio.Future] = None
+
+    async def user_getter():
+        nonlocal user_prom
+        if user_prom is None:
+            user_prom = asyncio.Future()
+            session_user = await get_user(session_recipe_user_id, user_context)
+            if session_user is None:
+                raise SuperTokensSessionError(
+                    SuperTokensSessionError.UNAUTHORISED, "Session user not found"
+                )
+            user_prom.set_result(session_user)
+        return await user_prom
+
+    mfa_requirements_for_auth = await Recipe.get_instance().recipe_implementation.get_mfa_requirements_for_auth(
+        {
+            "access_token_payload": access_token_payload,
+            "tenant_id": tenant_id,
+            "user": user_getter,
```

Comment:
this is incorrect translation, in Node, we have the user getter which is implicitly called when accessing the user property. Here however, it's a function. Is what we want even possible in python?

Reply:
Yes. Using task for that.